# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/webgpu/Device.cpp
# Commit: 41d7ed93df76
# Full Hash: 41d7ed93df76d84a8e842758f852d291a908e065
# Author: Brad Werth <bwerth@mozilla.com>
# Date: 2023-10-17 03:58:35
# Regressor Bug: 1838693
# File Overlap Count: 2
# Description:
#   Bug 1838693 Part 3: Make GPUDevice::Destroy() trigger wgpu device_destroy, then device_drop. r=webgpu-reviewers,nical
#   
#   This treats destroy as a 2-step process: wait on the destroy, then drop.
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D190238
# ==============================================================================

diff -r f214290956c1 -r 41d7ed93df76 dom/webgpu/Device.cpp
--- a/dom/webgpu/Device.cpp	Mon Oct 16 15:32:00 2023 +0000
+++ b/dom/webgpu/Device.cpp	Mon Oct 16 15:32:00 2023 +0000
@@ -97,6 +97,10 @@
                              aMessage);
 }
 
+void Device::TrackBuffer(Buffer* aBuffer) { mTrackedBuffers.Insert(aBuffer); }
+
+void Device::UntrackBuffer(Buffer* aBuffer) { mTrackedBuffers.Remove(aBuffer); }
+
 void Device::GetLabel(nsAString& aValue) const { aValue = mLabel; }
 void Device::SetLabel(const nsAString& aLabel) { mLabel = aLabel; }
 
@@ -358,7 +362,21 @@
 }
 
 void Device::Destroy() {
-  // TODO
+  if (IsLost()) {
+    return;
+  }
+
+  // Unmap all buffers from this device, as specified by
+  // https://gpuweb.github.io/gpuweb/#dom-gpudevice-destroy.
+  dom::AutoJSAPI jsapi;
+  if (jsapi.Init(GetOwnerGlobal())) {
+    IgnoredErrorResult rv;
+    for (const auto& buffer : mTrackedBuffers) {
+      buffer->Unmap(jsapi.cx(), rv);
+    }
+  }
+
+  mBridge->SendDeviceDestroy(mId);
 }
 
 void Device::PushErrorScope(const dom::GPUErrorFilter& aFilter) {